src/regex.rs

```diff
@@ -0,0 +1,82 @@
+use pgx::*;
+
+#[pg_schema]
+mod _prom_ext {
+
+    use pgx::*;
+    use regex::Regex;
+    use std::sync::Mutex;
+    use uluru::LRUCache;
+
+    // On caching: Creating a new Regex instance is expensive, so we keep a
+    // global cache of Regex instances in an LRU cache.
+    // An alternative approach would be to create a new regex type, and provide
+    // match functions between our custom regex type and TEXT (see pgpcre [1]).
+    // This would work, but requires dealing with the fact that somebody might
+    // decide to _store_ the compiled regex, which we don't want to support.
+    // [1]: https://github.com/petere/pgpcre/blob/master/pgpcre.c
+    // Interestingly, the Postgres' native regex engine is quite similar,
+    // storing compiled regex expressions in an LRU cache of 32 elements.
+
+    // On memory contexts: This cache and function do not behave nicely with
+    // Postgres' memory models. Allocations are not in a MemoryContext, instead
+    // they are directly on the stack, or heap. This is safe because we do not
+    // ever pass these objects to Postgres.
+
+    // Caveats: We completely ignore collation, and character sets.
+
+    struct CompiledRegex {
+        pattern: String,
+        matcher: Regex,
+    }
+
+    // Note: The chosen size is the same as Postgres' internal regex cache
+    const CACHE_SIZE: usize = 32;
+
+    lazy_static! {
+        static ref CACHE: Mutex<LRUCache<CompiledRegex, CACHE_SIZE>> =
+            Mutex::new(LRUCache::default());
+    }
+
+    /// re2_match matches `string` against `pattern` using an [RE2-like][re2]
+    /// regular expression engine, returning a `BOOLEAN`.
+    /// [re2]: https://github.com/google/re2
+    #[pg_extern(immutable, parallel_safe)]
+    fn re2_match(string: &str, pattern: &str) -> bool {
+        let mut cache = CACHE.lock().unwrap();
```

Comment:
It does lock the cache, yes. I'm not certain that it is a bottleneck. If the process which locks the cache would be multi-threaded, then this would become a contention point, and we would probably want to use a different access method.

Because Postgres' parallelism model is process-oriented, each process is actually operating on its own copy of the cache, so it's the only thing accessing the cache (I _think_). You may ask then why bother wrapping this in a Mutex at all? I believe that the answer is: Rust can't know that this is the case, and it wants a Mutex to be used. If all of the above is correct, then we _could_ use `unsafe` to bypass Rust and directly access the cache.

Reply:
> I'd argue the LRUCache is way too complicated for our usecase

Do you propose to not cache at all, or use a different caching mechanism?

Reply:
Different caching mechanism/library, as in our case we add a lot of machinery built to handle concurrent cache for single-threaded use case.
[This ](https://crates.io/crates/lru) looks promising.

Reply:
> we add a lot of machinery

Can you describe what you mean exactly?

The library that you linked seems to be very similar to the library that I've used (https://docs.rs/uluru/latest/uluru/). Is there something specific about that library which is better?

I have looked into bypassing the mutex and directly accessing the cache. It works, and is slightly faster. At the cost of using `unsafe`. The speed difference is quite small though.